assets/src/edit-story/app/story/effects/useSaveMetaBoxes.js

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { useEffect, useRef, useState } from 'react';
+import { useFeature } from 'flagged';
+
+/**
+ * Internal dependencies
+ */
+import { useAPI } from '../../api';
+
+/**
+ * Function returning the current Meta Boxes DOM Node in the editor
+ * whether the meta box area is opened or not.
+ * If the MetaBox Area is visible returns it, and returns the original container instead.
+ *
+ * @param {string} location Meta Box location.
+ * @return {Element} Container
+ */
+function getMetaBoxContainer(location) {
+  // Class name as set in <MetaBoxesArea>
+  const area = document.querySelector(
+    `.web-stories-meta-boxes-area-${location} .metabox-location-${location}`
+  );
+  if (area) {
+    return area;
+  }
+
+  return document.querySelector('#metaboxes .metabox-location-' + location);
+}
+
+function usePrevious(value) {
+  const ref = useRef();
+  useEffect(() => {
+    ref.current = value;
+  });
+  return ref.current;
+}
+
+function useSaveMetaBoxes({ story, isSaving, isAutoSaving }) {
+  const isFeatureEnabled = useFeature('customMetaBoxes');
+
+  const {
+    actions: { saveMetaBoxes },
+  } = useAPI();
+
+  const [isSavingMetaBoxes, setIsSavingMetaBoxes] = useState(false);
+
+  const wasSaving = usePrevious(isSaving);
+  const wasAutoSaving = usePrevious(isAutoSaving);
+
+  // Save metaboxes when performing a full save on the post.
+  // TODO: only do this when there are actually meta boxes.
+  useEffect(() => {
+    // Save metaboxes on save completion, except for autosaves that are not a post preview.
+    if (
+      !isFeatureEnabled ||
+      isSaving ||
+      isAutoSaving ||
+      isSavingMetaBoxes ||
+      !wasSaving ||
+      wasAutoSaving
+    ) {
+      return;
+    }
```

Comment:
This is key:

We only want to save meta boxes when we _just_ finished saving the story, but not if that save was an autosave.

Reply:
Should this only be triggered if there are any metaboxes to save? Also, is it always saving even if there are no actual changes? Could you confirm this is not influencing history?

Reply:
> Should this only be triggered if there are any metaboxes to save?
Good point, definitely should only trigger this if there are actually meta boxes. Will add this.

> Also, is it always saving even if there are no actual changes?
That's right, we don't (and can't really) detect whether there were any changes to the meta boxes. We just send the update and WordPress will figure it out server-side.

> Could you confirm this is not influencing history?

I'll double check, but this should have no impact on history at all, meta box changes are not added to history.

Reply:
Hmm, if it doesn't influence the history at all, then I wonder how can the user save only the updates from the meta-boxes in Draft mode? E.g. the user has no new changes and then changes the content of the meta boxes. Since history doesn't detect any new changes, the Draft button stays disabled. Thoughts?

Reply:
^ @swissspidy Thoughts on this in case you missed it among the notifications?

Reply:
So the way Gutenberg does it is by forcing  the "Save" button to be enabled if there are any active meta boxes:

https://github.com/WordPress/gutenberg/blob/f4300ff87ea9a8e5d2aed155f180c76b83a54c4a/packages/edit-post/src/components/header/index.js#L46-L58

That could be a viable option for us as well, because trying to add meta boxes to history or generally detecting any changes in meta boxes is going to be very complex and most probably not worth trying.

Reply:
Makes sense to me!